distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Period.enso

```diff
@@ -0,0 +1,176 @@
+
+    ## PRIVATE
+
+       Arguments:
+       - internal_period: An internal representation of period of type
+         java.time.Period.
+    Period_Without_Date internal_period
+
+    ## PRIVATE
+
+       Arguments:
+       - internal_period: An internal representation of period of type
+         java.time.Period.
+       - start_date: Starting Date
+    Period_With_Date internal_period start_date
+
+    ## Get the portion of the period expressed in years.
+    years : Integer
+    years self = self.internal_period.getYears
+
+    ## Get the portion of the period expressed in months.
+    months : Integer
+    months self = self.internal_period.getMonths
+
+    ## Get the portion of the period expressed in days.
+    days : Integer
+    days self = self.internal_period.getDays
+
+    ## Add the specified amount of time to this period.
+
+       Arguments:
+       - other_period: The period to add to `self`. Note that this cannot be a
+         `Duration`, neither `Date_Time`.
+
+       > Example
+       Add 1 day to 1 month.
+
+           import Standard.Base.Data.Time.Period
+
+           example_add = (Period.months 1) + (Period.days 1)
+    + : Period -> Period ! (Time_Error | Illegal_Argument_Error)
+    + self other_period =
+        ensure_period other_period <|
+            Panic.catch Any (Period.Period_Without_Date (self.internal_period.plus other_period.internal_period)) err->
+                case err of
+                    _ : DateTimeException -> Error.throw Time_Error_Data "Period addition failed:"+err.getMessage
+                    _ : ArithmeticException -> Error.throw Illegal_Argument_Error_Data "Arithmetic error:"+err.getMessage cause=err
+
+    ## Subtract a specified amount of time from this period.
+
+       Arguments:
+       - other_period: Other Period to add to this Period. Note that this
+         cannot be a `Duration`, neither `Date_Time`.
+
+       > Example
+         Subtract 11 months from a period of 3 years
+
+             import Standard.Base.Data.Time.Period
+
+             example_subtract = (Period.years 3) - (Period.months 11)
+
+    - : Period -> Period ! (Time_Error | Illegal_Argument_Error)
+    - self other_period =
+        ensure_period other_period <|
+            Panic.catch Any (Period.Period_Without_Date (self.internal_period.minus other_period.internal_period)) err->
+                case err of
+                    DateTimeException -> Error.throw Time_Error_Data "Period subtraction failed"
+                    ArithmeticException -> Error.throw Illegal_Argument_Error_Data "Arithmetic error"
+
+    ## Check two periods for equality.
+
+       Note that two periods are equal if they have the exact same amount of
+       years, months, and days. So `(Period.days 30)` and
+       `(Period.new months=1`) are not equal.
+
+       Arguments:
+       - other_period: The period to compare against `self`.
+    == : Period -> Boolean ! (Time_Error | Illegal_Argument_Error)
+    == self that =
```

Comment:
I don't think we should rely on the `compare_to` for this.
Ideally I want `Period.new years=1 == Period.new months=12`

Reply:
Think should be `self.total_months == other.total_months && self.days == other.days`

Reply:
> Ideally I want Period.new years=1 == Period.new months=12

12 months may not be equal to a year. An expression `Period.years 1 == Period.months 12` will always be true. However, such an expression will rarely be used in practice. In practice, users will get `Period` most likely from `Period.between`, and if 12 months is computed for a leap year, and 1 year for a standard year, than `12 months > 1 year`.

Reply:
A period should always be operating on a calendar basis, duration is a time based.
So regardless of when it starts adding months/years should just alter the month and year part.
The only tuning is what happens if the day doesn't exist in the next month (e.g. 31st going to shorter month), but we will just trust the Java handling there (31st August plus 1 month == 30th September).

Reply:
Since we agreed not to implement `compare_to` at all, let's implement this as `self.days == other.days && self.months == other.months && self.years == other.years`, maybe with attaching a warning? Or do we rather want `Period.==` to just throw `Incomparable_Values_Error`?

Reply:
That is consistent with Java but still feel 12 month should equal 1 year but happy otherwise.

Reply:
OK, let's be consistent with java then and compare it field-by-field.

Reply:
I also think that `12 months` == `1 year` makes sense, there's no imprecision here.

If we go by the above example with leap years - we couldn't even compare `1 year == 1 year` (because one of the years could have been leap and other not) or `1 month == 1 month` because one of the months could have been July and the other February and so their day count would differ. But I think at this point we are not comparing day count, but just if its the same _"logical"_ period - and while one cannot judge if `1 month` and `30 days` represent the same logical period (as this depends on the context), IMO `12 months` and `1 year` always do represent the same logical period.